eox_tenant/signals.py

```diff
@@ -217,3 +222,114 @@ def clear_tenant(sender, request, **kwargs):  # pylint: disable=unused-argument
     Signal: django.core.signals.got_request_exception
     """
     pass
+
+
+def tenant_context_addition(sender, body, headers, *args, **kwargs):  # pylint: disable=unused-argument
+    """
+    Receiver used to add context to the async process about the tenant associated to the task that need to be done.
+
+    Dispatched before a celery task is published. Note that this is executed in the process sending the task.
+    See:
+        https://celery.readthedocs.io/en/latest/userguide/signals.html#before-task-publish
+    """
+    get_host_func = AsyncTaskHandler().get_host_from_task(sender)
+    headers['eox_tenant_sender'] = get_host_func(body)
+
+
+def start_async_tenant(sender, *args, **kwargs):  # pylint: disable=unused-argument
+    """
+    Receiver that runs on the async process to update the settings accordingly to the tenant.
+    Dispatched before a task is executed.
+    See:
+       https://celery.readthedocs.io/en/latest/userguide/signals.html#task-prerun
+    """
+    headers = sender.request.get('headers') or {}
+    http_host = headers.get('eox_tenant_sender')
+    if not http_host:
+        LOG.warning("Could not find the host information for eox_tenant.signals. ")
+        siteid = base_settings.SITE_ID
+        site = Site.objects.get(id=siteid)
+        http_host = site.name
+
+    try:
+        domain = http_host.split(':')[0]
+    except KeyError:
+        siteid = base_settings.SITE_ID
+        site = Site.objects.get(id=siteid)
+        http_host = site.name
+        domain = http_host.split(':')[0]
+
+    if reset_or_keep_settings(domain):
+        return
+
+    # Do the update
+    _update_settings(domain)
+
+
+class AsyncTaskHandler(object):
+    """
+    Handler used to get the tenant of an async task.
+    """
+
+    sender = None
+
+    def get_host_from_task(self, sender):
+        """
+        Method used to get the function used to get the tenant domain
+        accordingly to the type of task.
+        """
+        self.sender = sender
+        value = base_settings.EOX_TENANT_ASYNC_TASKS_HANDLER_DICT.get(sender, 'default_host')
+        action = getattr(self, value)
+        return action()
+
+    def get_host_from_siteid(self):
+        """
+        Get the tenant domain name using its Site id
+        """
+        def get_host(body):
+            """
+            Get host from siteid.
+            """
+            try:
+                siteid = body.get('args')[0]
+                site = Site.objects.get(id=siteid)
+                domain = site.name
+            except (Site.DoesNotExist, KeyError, TypeError):  # pylint: disable=no-member
+                return self.default_host()(body)
+            return domain
+
+        return get_host
+
+    def default_host(self):
+        """
+        Get default tenant.
+        """
+        LOG.warning(
```

Comment:
If this is the default way of getting the context, I do not think that such a warning is required just yet.

First we try to find if we are in a sync process that has already an active tenant. If so we use that tenant.
Then if no tenant is available we log the error and as much information as we can. E.g The traceback.